server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java

```diff
@@ -101,47 +92,38 @@ public void configure(HierarchicalConfiguration<ImmutableNode> config) {
     @PostConstruct
     public void init() {
         LOGGER.info("init...");
-        queue = queueFactory.createQueue(MailQueueFactory.SPOOL);
-        spooler = Schedulers.fromExecutor(Executors.newFixedThreadPool(numThreads, NamedThreadFactory.withName("spooler")));
-        LOGGER.info("uses {} Thread(s)", numThreads);
-        run();
+        LOGGER.info("Concurrency level is {}", concurrencyLevel);
+        queue = queueFactory.createQueue(MailQueueFactory.SPOOL, MailQueueFactory.prefetchCount(concurrencyLevel));
+        disposable = run(queue);
+        LOGGER.info("Spooler started");
     }
 
-    private void run() {
-        LOGGER.info("Queue={}", queue);
-        disposable = Flux.from(queue.deQueue())
-            .flatMap(item -> handleOnQueueItem(item).subscribeOn(spooler), parallelismLevel)
+    private reactor.core.Disposable run(MailQueue queue) {
+        return Flux.from(queue.deQueue())
+            .flatMap(item -> handleOnQueueItem(item).subscribeOn(Schedulers.elastic()), concurrencyLevel)
             .onErrorContinue((throwable, item) -> LOGGER.error("Exception processing mail while spooling {}", item, throwable))
-            .subscribeOn(spooler)
+            .subscribeOn(Schedulers.elastic())
             .subscribe();
     }
 
     private Mono<Void> handleOnQueueItem(MailQueueItem queueItem) {
-        TimeMetric timeMetric = metricFactory.timer(SPOOL_PROCESSING);
-        try {
-            return Mono.fromCallable(processingActive::incrementAndGet)
-                .flatMap(ignore -> processMail(queueItem).subscribeOn(spooler))
-                .doOnSuccess(any -> timeMetric.stopAndPublish())
-                .doOnSuccess(any -> processingActive.decrementAndGet());
-        } catch (Throwable e) {
-            return Mono.error(e);
-        }
+        return metricFactory
+            .runPublishingTimerMetric(SPOOL_PROCESSING,
+                 Mono.fromCallable(processingActive::incrementAndGet)
+                    .flatMap(ignore -> processMail(queueItem))
+                    .doOnSuccess(any -> processingActive.decrementAndGet()));
```

Comment:
Not related, but this is not decremented if processing fails?

Reply:
yes, good catch, will fix